### DISCUSSION TITLE
How can I embed an already serialized object in another?

### DISCUSSION THREAD

**Discussion by wuyuanyi135** [nlohmann/json#3967] [Q&A]
Hi, I am trying to achieve something like:
```cpp
  json internal_obj;
  internal_obj["time"] = {{"from", 1},
                          {"to",   100}};
// The internal buffer is stored somewhere
  std::vector<uint8_t> internal_buffer;
  json::to_cbor(internal_obj, internal_buffer);

// Later, the internal buffer will be loaded and embedded as-is into the ext_obj and send.
// I want to avoid the overhead of repeating the decode/encode process.
  json ext_obj;
  ext_obj["logging"] = json::binary(internal_buffer);
  std::vector<uint8_t> out{};
  json::to_cbor(ext_obj, out);

  std::for_each(out.begin(), out.end(),[](const auto& c){
    fmt::print("{:02x}", c);
  });
```

I could embed the serialized object as a `binary` part in cbor (or `string` part in JSON). However, to retrieve the value, the receiver must recursively decode the binary inside it. Is there any recommended way so that I can embed the serialized object as-is, and modify its type for dumping?

**Comment by nlohmann** [comment]
There's no API to splice an already-serialized CBOR/MessagePack byte buffer into another json value's tree without decoding it — json::binary() only stores an opaque byte blob (it doesn't know it's CBOR-encoded), so a receiver would indeed have to know to treat that particular binary field as embedded CBOR and decode it separately (via json::from_cbor on the extracted bytes) rather than it happening implicitly during the outer from_cbor call. If you want to avoid the decode/re-encode round trip for the outer object but still get one flat structure, you'd have to write your own recursive binary-parsing/embedding step outside the library; there's no built-in 'raw embed and later flatten' feature.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
